### PR TITLE
Add a common subclass for IndexNode::iterator

### PR DESCRIPTION
**Pull request by zhengbuqian**
so that different indexes only need to implement next_batch. Any index that gradually adds new points as the `iterator::Next()` is called should implement their iterator by inheriting this new class.

Updated the impl for supported ivf and hnsw index.

issue: https://github.com/zilliztech/knowhere/issues/500
/kind: enhancement

### CODE REVIEW

File: include/knowhere/index_node.h
Comment (alexanderguzhva) [inline include/knowhere/index_node.h:198]: should be `float`

File: include/knowhere/index_node.h
Comment (alexanderguzhva) [inline include/knowhere/index_node.h:201]: there are too many `std::pair<float, int64_t>` structs in the code. I'd create a simple 2-field structure with clear field names, because the current code uses `.first` or `.second` and who knows that they mean

File: src/index/ivf/ivf.cc
Comment (alexanderguzhva) [inline src/index/ivf/ivf.cc:251]: you can make this function `constexpr`
Also, what about SCANN?

**Comment by zhengbuqian** [comment]
/hold

this branch is currently based on #505, need to rebase after #505 is merged

File: include/knowhere/index_node.h
Comment (zhengbuqian) [reply include/knowhere/index_node.h:138]: initialization must be done after constructor of subclass has finished, which we don't know when that'll happen in this class

**Review by foxspy — APPROVED** [review]
/lgtm

**Pull request merged**
